### PR TITLE
Extract common functionality of iOS platformviews into superclasses

### PR DESCRIPTION
**Pull request by yaakovschectman**
Move most functionality of `UiKitView` and its supporting classes into superclasses named `DarwinPlatformView`, etc., and create trivial or near-trivial subclasses with the same names as the old classes.

I am currently awaiting approval for a macOS workstation that would allow me to run the iOS/macOS tests and make sure all existing functionality is preserved by this refactor. I can ensure that tests will pass, but doing so may need to wait for a while.

Addresses [Add AppKitView](https://github.com/flutter/flutter/issues/128519)

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (loic-sharma) [inline packages/flutter/lib/src/rendering/platform_view.dart:272]:
```suggestion
/// Superclass for common functionality of platform view RenderBoxes
```

Comment (goderbauer) [inline packages/flutter/lib/src/rendering/platform_view.dart:272]: Can you add a little more color to this doc? See also our style guide for writing prompts: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#writing-prompts-for-good-documentation

Questions to answer: 
* What is the common functionality that this provides?
* What responsibilities are left to the subclasses?
* Can this class be use directly? (it's documented as a super class, but it isn't abstract, so what do I get if I instantiate it directly?)
* Document what the generic type parameter `T` is

Comment (goderbauer) [reply packages/flutter/lib/src/rendering/platform_view.dart:272]: nit: doc sentences should be grammatically correct and end with a `.`, https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#use-correct-grammar

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/platform_view.dart:276]: You could also just remove this sentence entirely since we no longer document that something must not be null now that we have full null safety and the types are self-documenting.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/platform_view.dart:272]: > What responsibilities are left to the subclasses?

Looks like at least gesture handling needs to be provided by the sub class?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/platform_view.dart:580]: this line should probably be reformatted into multi lines to make it more readable...

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/platform_view.dart:762]: should this get a `@mustCallSuper`?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/platform_view.dart:764]: Convention is to use cascades here:

```suggestion
    renderObject
      ..viewController = controller
      ..hitTestBehavior = hitTestBehavior;
```

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [inline packages/flutter/lib/src/widgets/platform_view.dart:785]: nit: remove unnecessary blank line.

**Review by goderbauer — COMMENTED** [review]
For my own understanding: The motivation for this refactoring is to add a AppKitPlatformView later that will reuse the DarwinPlatformView machinery?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/widgets/platform_view.dart:764]: Do you think changes such as this to the code that was left unchanged aside from renaming the enclosing class, such as this, should be within the scope of this PR?

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/rendering/platform_view.dart:273]: I do not think it can be private as it needs to be referenced from another source file as well.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/rendering/platform_view.dart:404]: @cbracken and I discussed this a little as well. I believe our approach to gesture handling will need to be decoupled between iOS and macOS. Please feel free to clarify further if you remember.

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/platform_view.dart:764]: For me, tiny clean-ups like this are always in scope, especially if you touch the code anyways like here.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (goderbauer) [reply packages/flutter/lib/src/rendering/platform_view.dart:404]: Would it make sense to wait with this refactoring then until we have an AppKit implementation and have clarity on what's common and what's not?

File: packages/flutter/lib/src/services/platform_views.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/services/platform_views.dart:1387]: This seemed useful in case we need to make decoupling changes between iOS and macOS, especially as the macOS implementation is still planned.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/rendering/platform_view.dart:404]: As I understand, implementing AppKit would require either already having superclasses with common functionality in place, or duplicating everything that they would have in common, which was the impetus for this refactor.

Comment (cbracken) [reply packages/flutter/lib/src/rendering/platform_view.dart:405]: Gesture handling is not yet supported on macOS, but will likely be within the next week or two. Given that, perhaps we should push this up to the superclass now rather than later. API-wise this is pretty safe to support even before we actually have the embedder side work done I think. (We're cheating and testing by using UIKitView already as-is).

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (cbracken) [inline packages/flutter/lib/src/rendering/platform_view.dart:275]: Format nit: Extract out a summary sentence and break it into a separate paragraph. https://dart.dev/effective-dart/documentation#do-separate-the-first-sentence-of-a-doc-comment-into-its-own-paragraph

I don't think we ever explicitly say to avoid using "superclass for" but I wonder if we can just shorten this to:

```dart
/// Common render-layer functionality for iOS and macOS platform views.
///
/// Further details in follow-on paragraph here.
```

**Comment by yaakovschectman** [comment]
@cbracken `perhaps we should push this up to the superclass now rather than later`
Because the constructor for the gesture recognizer takes a `UiKitViewController` parameter, this would require either all `RenderDarwinPlatformView` to hold a `UiKitViewController`, or update `_UiKitViewGestureRecognizer` to take the superclass `DarwinPlatformViewController` instead.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (cbracken) [reply packages/flutter/lib/src/services/platform_views.dart:1387]: We need it for backward compatibility since these classes have existed for a long time on iOS. Longer term, it may well turn out we can merge all of this into a single class, but that'll require a deprecation process.

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (cbracken) [inline packages/flutter/lib/src/rendering/platform_view.dart:285]:
```suggestion
  /// The unique identifier of the platform view controlled by this controller.
```

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (cbracken) [inline packages/flutter/lib/src/rendering/platform_view.dart:384]:
```suggestion
/// UIViews are added as subviews of the FlutterView and are composited by Quartz.
```

File: packages/flutter/lib/src/services/platform_views.dart
Comment (cbracken) [inline packages/flutter/lib/src/services/platform_views.dart:1316]: Perhaps something along the lines of:
```suggestion
/// Base class for iOS and macOS view controllers.
///
/// View controllers are used to create and interact with the UIView or NSView
/// underlying a platform view.
```

File: packages/flutter/lib/src/services/platform_views.dart
Comment (cbracken) [inline packages/flutter/lib/src/services/platform_views.dart:1386]: Similar comment here. This could probably be fleshed out a little.

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (cbracken) [inline packages/flutter/lib/src/widgets/platform_view.dart:198]: Perhaps something like:
```suggestion
/// Common superclass for iOS and macOS platform views.
///
/// Platform views are used to embed native views in the widget hierarchy, with
/// support for transforms, clips, and opacity similar to any other Flutter widget.
```

**Review by cbracken — COMMENTED** [review]
Thanks for putting this together. 

Added a few suggestions on improving the doc comments, which admittedly were pretty bare-bones in the original code, but now is a great time to improve them, even just a little bit.

I'll give some thought as to whether there's a way we could avoid the generic parameters. The implications weren't super clear to me when first considering the problem. On the plus side, this is primarily an issue for private classes, but I'm starting to wonder if an alternative might be to use composition rather than inheritance here and just have both the UiKitView and AppKitView classes delegate to a common implementation class. What do you think?

**Comment by yaakovschectman** [comment]
If, for example, a `UIView` needs to access a method that only the iOS version of the view controller provides, wouldn't we still need generics or type casting even if using composition? Unless dart's covariance is more robust than I thought, in which case maybe overriding a getter could work?

File: packages/flutter/lib/src/rendering/platform_view.dart
Comment (hellohuanlin) [reply packages/flutter/lib/src/rendering/platform_view.dart:410]: I agree that it's better to push gesture stuff to the superclass. Or if you want to keep it here, add a TODO, and also a comment explaining why here.

File: packages/flutter/lib/src/services/platform_views.dart
Comment (hellohuanlin) [reply packages/flutter/lib/src/services/platform_views.dart:1320]: should be abstract since it's not supposed to be instantiated

File: packages/flutter/lib/src/services/platform_views.dart
Comment (hellohuanlin) [reply packages/flutter/lib/src/services/platform_views.dart:1392]: I'm wondering if there is a `typealias` in dart? 

`typealias UiKitViewController = DarwinPlatformViewController`

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (hellohuanlin) [inline packages/flutter/lib/src/widgets/platform_view.dart:202]: `_DarwinPlatformView`?

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (hellohuanlin) [reply packages/flutter/lib/src/widgets/platform_view.dart:596]: yeah `TController` is odd to me. Either use short form or long form in generic, but I've never seen a mix of the 2

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (hellohuanlin) [inline packages/flutter/lib/src/widgets/platform_view.dart:620]: why is it called private view?

**Review by hellohuanlin — APPROVED** [review]
basic idea lgtm after addressing all those comments.

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/widgets/platform_view.dart:620]: This template class corresponds to the private view class used by the subclass, e.g. `_UiKitPlatformView`, as opposed to the public class `UiKitView`.

**Comment by hellohuanlin** [comment]
> I'm starting to wonder if an alternative might be to use composition rather than inheritance here and just have both the UiKitView and AppKitView classes delegate to a common implementation class. 

@cbracken I have the same feeling when reviewing the code - inheritance makes it harder to follow the logic, as you have to jump between superclasses and subclasses. Composition better encapsulates the logic here and i think it's worth a try.

**Comment by yaakovschectman** [comment]
> Composition better encapsulates the logic here and i think it's worth a try.

I'm a little unclear how we would go about solving this with just composition, as each of the classes touched by this PR so far would need iOS and mac versions which do mostly the same things, but possibly with a few changes. I'm not really clear on how we could leverage composition to avoid code duplication while also allowing changes to one platform that don't affect the other.

**Comment by wanjm** [comment]
when will it be merged?

**Comment by yaakovschectman** [comment]
> when will it be merged?

After the reviewers have responded to my follow-ups to their comments. One reviewer who had questions will be out for another week or so.

@hellohuanlin If switching to composition is something you need addressed before this PR is merged, please address my above concerns.

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/widgets/platform_view.dart:620]: Looks like Google's style convention, at least in Java, is to append T at the end of long form generic types. I will update this

File: packages/flutter/lib/src/widgets/platform_view.dart
Comment (yaakovschectman) [reply packages/flutter/lib/src/widgets/platform_view.dart:202]: That name is already used by another class: the subclass of `LeadRenderObjectWidget`

Comment (yaakovschectman) [reply packages/flutter/lib/src/widgets/platform_view.dart:202]: > should be abstract since it's not supposed to be instantiated

We can either make it abstract, or we can typedef `UiKitViewController` to `DarwinPlatformViewController`. We could not do both. I will go with the former for now, as it will make adding iOS specific functionality easier.

**Review by cbracken — APPROVED** [review]
lgtm.

Summing up a discussion over chat re: composition vs inheritance. API-wise, a composition-based approach affords a tiny bit more future wiggle room should UIKit and AppKit based subclasses ever significantly diverge, since it's (relatively) easy to add a superclass that wasn't there, but (relatively) breaking to remove one that was. Further, composition avoids inserting a superclass that doesn't serve a significant conceptual purpose API-wise (users are likely to assign UIViews to PlatformViews but not likely to need a DarwinView specifically. However, composition involves duplicating a ton of API (across all the layers of the framework) and having dozens of calls out to some common private delegate. Overall, we think it's very unlikely that our UIKit and AppKit platformview classes diverge very significantly, and inheritance save a significant amount of code.

**Comment by Hixie** [comment]
test-exempt: code refactor with no semantic change

**Pull request merged**
